plasmapy/particles/particle_class.py

```diff
@@ -96,6 +104,91 @@ def _category_errmsg(particle, category: str) -> str:
     return errmsg
 
 
+class ParticleJSONDecoder(json.JSONDecoder):
+    """
+    A custom `JSONDecoder` class to deserialize particles.
+    Inherits from `json.JSONDecoder`. Refer to documentation at:
+    https://docs.python.org/3/library/json.html#json.JSONDecoder
+
+    Parameters
+    ----------
+    object_hook
+        operation to be performed to deserialize the
+        JSON string to a `dict`. The ParticleJSONDecoder uses a custom hook
+        called `particle_hook` to deserialize particle JSON strings.
+
+    parse_float
+        Specifies how to handle floating point numbers
+        in a JSON string. Refer to documentation link above for more details.
+
+    parse_int
+        Specifies how to handle integer numbers.
+        Refer to documentation link above for more details.
+
+    parse_constant
+        Specifies how to handle constants.
+        Refer to documentation link above for more details.
+
+    strict
+        Refer to documentation link above for details.
+
+    object_pairs_hook
+        Refer to documentation link above for details.
+    """
+
+    def __init__(
+        self,
+        *,
+        object_hook=None,
+        parse_float=None,
+        parse_int=None,
+        parse_constant=None,
+        strict=True,
+        object_pairs_hook=None,
+    ):
+
+        if object_hook is None:
+            object_hook = self.particle_hook
+
+        json.JSONDecoder.__init__(
+            self,
+            object_hook=object_hook,
+            parse_float=parse_float,
+            parse_int=parse_int,
+            parse_constant=parse_constant,
+            strict=strict,
+            object_pairs_hook=object_pairs_hook,
+        )
+
+    @staticmethod
+    def particle_hook(json_dict):
+        """
+        An `object_hook` utilized by the `json` deserialization processes to decode
+        json strings into a `plasmapy` particle class (`AbstractParticle`,
+        `CustomParticle`, `DimensionlessParticle`, `Particle`).
+        """
+        particle_types = {
+            "AbstractParticle": AbstractParticle,
+            "CustomParticle": CustomParticle,
+            "DimensionlessParticle": DimensionlessParticle,
+            "Particle": Particle,
+        }
+        if "plasmapy_particle" in json_dict:
+            try:
+                pardict = json_dict["plasmapy_particle"]
+                partype = pardict["type"]
+                args = pardict["__init__"]["args"]
+                kwargs = pardict["__init__"]["kwargs"]
+                particle = particle_types[partype](*args, **kwargs)
+                return particle
+            except KeyError:
+                raise InvalidElementError(
+                    f"json file does not define a valid plasmapy particle"
+                )
+        else:
+            return json_dict
```

Comment:
I think we could extract this whole staticmethod to be a function, and then we get a programmatic way of converting a dictionary into a Particle for free! Then we could simply have this method be a thin wrapper on top of that function.

Reply:
I debated this myself.  I had if left like this because it's consistent with the `JSONDecoder` framework.  We could drop the `JOSNDeocder` for something more like...

```python
def particle_hook(....):
    ...

with open('particle.json', 'r') as fp:
    par = json.load(fp, object_hook=particle_hook, ...)
```

I could go either way on this, but I think I slightly prefer the `JSONDecoder` framework.   In the end, I say we leave the PR as it is and, if the need arises to convert dictionaries into particles, then we can modify with a separate PR.